lecture_3/template/nkt2111.tex

```diff
@@ -0,0 +1,255 @@
+
+\begin{enumerate}
+    \item $T(n)=a_0 + a_1n + a_2 n^2 + \ldots + a_dn^d = \mathcal{O}(n^d) = \Omega(n^d) = \Theta(n^d)$
+    \item $\mathcal{O}(\log_a n) = \mathcal{O}(\log_b n)$ for $a, b >0$ (base does not matter, it is another constant)
+    \item $n^d = \mathcal{O}(r^n)$ $\forall$ $r>1, d>0$
+\end{enumerate}
+
+\subsection{Common Running Times}
+Next, Sid walked through some common running times and gave examples of algorithms and data structures where these running times are commonly seen. It is important to note that here, we will assume that arithmetic, indexing, and "basic operations" take $\mathcal{O}(1)$ time. 
+
+It is also essential to note something that has been absent from our discussion up to this point, but which is essential to keep in mind. That is thinking about the amount of space we take up when implementing certain data structures or algorithms. There is often a trade off between space and time, and space is important to consider because even if an algorithm operates in sublinear time, it might require storage of massive amounts of data, which could be inefficient or even impossible. Here, we will proceed with some common running times but keep the space complexity in the back of our minds as well.
+
+\begin{enumerate}
+    
+    \item $\mathcal{O}(\log(n))$ \\
+    Logarithmic run time is often seen when working with binary search trees, which is a commonly used data structure that has an inherent sorting structure which makes it so we do not have to visit every node. For example, inserting or searching in a BST has logarithmic run time. \\
+    
+    \begin{figure}[h!]
+  \begin{center}
+    \includegraphics[width=0.4\textwidth]{figures/BSTSearch.png}.
+    \caption{
+      Binary Search Tree Visual (from https://www.geeksforgeeks.org/binary-search-tree-data-structure/)}
+    \label{fig:example_figure}
+  \end{center}
+\end{figure}
+    
+    Another data structure whose associated algorithms often have this run time is a skip list. Sid noted that we can think of the skip list as a sort of multi-leveled linked list. If the bottom level has n nodes, the next level up will have $\frac{n}{2}$ nodes and the next level after that will have $\frac{n}{4}$ nodes. The pattern continues so that each subsequent layer has half as many nodes as the previous layer. This data structure makes searching more efficient because it is possible to start at the level that is farthest out, which is analogous to starting at the first/root node in the BST. Thus, the run time for searching, similar to for a BST is $\mathcal{O}(\log(n))$. \\
+    
+    \begin{figure}[h!]
+  \begin{center}
+    \includegraphics[width=0.5\textwidth]{figures/A-skip-list-is-a-set-of-linked-lists-Numbers-within-17-are-the-indexes-of-the.png}.
+    \caption{
+      Skip List Visual (from https://www.researchgate.net/figure/A-skip-list-is-a-set-of-linked-lists-Numbers-within-17-are-the-indexes-of-the_fig2_221560836)}
```

Comment:
this breaks because of the underscores in the file name---latex thinks your trying to use '_' as a subscript in math mode.

wrap the whole url in \url{ ... }, which will also make it clickable in the pdf.